apps/verticalhome/js/pin_app_navigation.js

```diff
@@ -0,0 +1,294 @@
+    },
+
+
+    /**
+     * Handle event keydown
+     * @param  {[Object]} e [default object]
+     * @return {[nothing]}   [nothing]
+     */
+    handleEvent: function(e) {
+
+      function cycleElements(index) {
+
+        var removed = this.elements.splice(-index, index);
+
+        for (var i = 0; i < removed.length; i++) {
+
+          this.elemList.insertBefore(removed[i], this.elements[0 + i]);
+          this.elements.splice(0 + i, 0, removed[i]);
+
+        }
+
+      };
+
+      if (this.preventDef) {
+        e.preventDefault();
+      }
+      
+      function preVerticalNavigation () {
+        if (this.middleElem) {
+          document.querySelector('.middle').classList.remove('middle');
+          this.middleElem = false;
+        }
+
+
+        var toRemoved = this.elemList.querySelector('.removed');
+        var toRemoveOutFocus = this.elemList.querySelector('.out-focus');
+
+        if (toRemoved) {
+          this.elemList.removeChild(toRemoved);
+        }
+
+        if (toRemoveOutFocus) {
+          toRemoveOutFocus.classList.remove('out-focus');
+        }
+      };
+
+      if (this.preventDef){
+        e.preventDefault();
+      }
+
+
+      switch(e.key) {
+        case 'ArrowUp':
+
+          preVerticalNavigation.call(this);
+          document.getElementById('clock').style.opacity = 0;
+
+          if (this.elemList.dataset.scrollup) {
+            this.elemList.style.top = this.elemList.dataset.scrollup;
+          } else {
+            this.elemList.style.top = '-6rem';
+            this.elemList.dataset.scrollup = this.elemList.style.top;
+          }
+
+          var deltaElemsToStartScroll = 3;
+
+          this.elements[this.selectedElemIndex].classList.remove('selected');
+          if (this.selectedElemIndex >= 0 && this.selectedElemIndex < 4) {
+
+            cycleElements.call(this, (deltaElemsToStartScroll - this.selectedElemIndex));
+            this.selectedElemIndex += deltaElemsToStartScroll - this.selectedElemIndex;
+
+            this.selectedElemIndex--;
+            this.elements[this.selectedElemIndex].classList.add('selected');
+
+          } else {
+
+            this.selectedElemIndex--;
+            this.elements[this.selectedElemIndex].classList.add('selected');
+
+            if (this.selectedElemIndex > 3) {
+              this.elemList.style.top = parseFloat(this.elemList.style.top) + 6 + 'rem';
+              this.elemList.dataset.scrollup = this.elemList.style.top;
+            }
+
+          }
+
+          var lastElem = this.elements[this.elements.length - 1];
+          var clonedElem = lastElem.cloneNode(true);
+
+
+          this.elemList.insertBefore(clonedElem, this.elements[0]);
+          this.elements.splice(0, 0, clonedElem);
+          this.elements[0].classList.add('out-focus');
+          this.selectedElemIndex++;
+
+          lastElem.classList.add('removed');
+          this.elements.splice(-1, 1);
+
+          this.prevVerticalKey = 'up';
+
+
+          break;
```

Comment:
That is a long case block. You should break it out into a function or method.

Also, the code is pretty much incomprehensible. This is some of the most important code for this homescreen app. It needs to be clear and maintainable.  I suspect it needs to be completely rewritten. But if really needs to be like this, then document it carefully and explain why you're cycling elements and cloning nodes and modifying the position of the container element.  It may be helpful to refer to the UX spec in your comments.

Does the UX spec specify that when the user reaches the end of the list the we cycle back around to the top?  If the code is that complex because of that requirement, consider implementing something simpler to start without that cycling around feature.

I've got the same genral comment for the arrowdown case below, of course.